### PR TITLE
[server/generate_once] write sessionData caches to file

### PR DESCRIPTION
**Pull request by grqz**
[server/generate_once] Cache sessionData to file
    - [server/generate_once] Better logging
    - [server/generate_once] Parse date with new Date()
    - [server/session_manager] add exporting interface YoutubeSessionDataCaches
    - [server/session_manager] add private property TOKEN_TTL_HOURS to SessionManager
    - [server/session_manager] add public method cleanupCaches, getYoutubeSessionDataCaches(boolean), setYoutubeSessionDataCaches(YoutubeSessionDataCaches)
    - [server/session_manager+generate_once] automatically cleanup caches
    - [server/session_manager] rename private property youtubeSessionData to youtubeSessionDataCaches

### CODE REVIEW

File: server/src/generate_once.ts [hunk 1]
```diff
@@ -14,37 +17,59 @@ const options = program.opts();
...
+            if (
+                parsedCaches[visitIdentifier] &&
+                typeof parsedCaches[visitIdentifier].generatedAt === "string"
+            ) {
+                const parsedDate = parsedCaches[visitIdentifier].generatedAt;
+                parsedCaches[visitIdentifier].generatedAt = new Date(
+                    parsedDate,
+                );
+            }
+        }
```
Comment (Brainicism) [inline server/src/generate_once.ts:35]:
```suggestion
            if (
                parsedCaches[visitIdentifier]
            ) {
                const parsedDate = parsedCaches[visitIdentifier].generatedAt as string;
                parsedCaches[visitIdentifier].generatedAt = new Date(
                    parsedDate,
                );
            }
        }
```

Comment (Brainicism) [reply server/src/generate_once.ts:35]: If you were just trying to tell the compiler that `generatedAt` is a string

File: server/src/generate_once.ts [hunk 2]
```diff
@@ -14,37 +17,59 @@ const options = program.opts();
...
     } else {
-        if (verbose) {
-            console.log(
-                `Received request for visitor data, grabbing from Innertube`,
-            );
-        }
+        log(`Received request for visitor data, grabbing from Innertube`);
         const generatedVisitorData = await sessionManager.generateVisitorData();
         if (!generatedVisitorData) {
+            // Should we remove this console.error? There's one already in sessionManager.generateVisitorData().
```
Comment (Brainicism) [inline server/src/generate_once.ts:57]: sure

File: server/src/generate_once.ts [hunk 3]
```diff
@@ -1,6 +1,9 @@
-import { SessionManager } from "./session_manager";
+import { SessionManager, YoutubeSessionDataCaches } from "./session_manager";
 import { Command } from "@commander-js/extra-typings";
+import * as fs from "fs";
+import * as path from "path";
 
+const CACHE_PATH = path.resolve(process.cwd(), "cache.json");
```
Comment (Brainicism) [inline server/src/generate_once.ts:6]: Can you confirm what `process.cwd()` resolves to? I'm assuming its the path of the python plugin that invokes it?

**Review by Brainicism — CHANGES_REQUESTED** [review]

File: server/src/generate_once.ts (hunk 3 above)
Comment (Brainicism) [reply server/src/generate_once.ts:6]: If we want to make it agnostic of the process's current working directory, we can use a path relative to `__dirname` (the directory of the current file)

Comment (grqz) [reply server/src/generate_once.ts:6]: it's where you run yt-dlp

File: server/src/generate_once.ts [hunk 4]
```diff
@@ -10,41 +12,58 @@ program.parse();
 const options = program.opts();
 
 (async () => {
+    const CACHE_PATH = path.resolve(__dirname, "cache.json");
```
Comment (grqz) [inline server/src/generate_once.ts:15]:
```suggestion
    const CACHE_PATH = path.resolve(__dirname, "..", "cache.json");
```
do you think it should be in the same dir with `package.json`?

Comment (Brainicism) [reply server/src/generate_once.ts]: i think so

File: server/src/generate_once.ts [hunk 5]
```diff
@@ -1,6 +1,9 @@
-import { SessionManager } from "./session_manager";
+import { SessionManager, YoutubeSessionDataCaches } from "./session_manager";
 import { Command } from "@commander-js/extra-typings";
+import * as fs from "fs";
+import * as path from "path";
 
+const CACHE_PATH = path.resolve(__dirname, "cache.json");
```
Comment (grqz) [inline server/src/generate_once.ts:6]:
```suggestion
const CACHE_PATH = path.resolve(__dirname, "..", "cache.json");
```

File: server/src/generate_once.ts [hunk 6]
```diff
@@ -14,37 +17,56 @@ const options = program.opts();
...
+                        poToken: parsedCache.poToken,
+                        generatedAt: new Date(parsedCache.generatedAt),
+                        visitIdentifier,
+                    };
+                }
+            }
+            cache = parsedCaches;
+        } catch (e) {
+            log(`Error parsing cache. e = ${e}`);
+            cache = {};
```
Comment (grqz) [inline server/src/generate_once.ts:39]:
```suggestion
```
already assigned `{}` at line 20, don't need this

File: server/src/generate_once.ts [hunk 7]
```diff
@@ -14,37 +17,56 @@ const options = program.opts();
...
+                const parsedCache = parsedCaches[visitIdentifier];
+                if (parsedCache) {
+                    cache[visitIdentifier] = {
+                        poToken: parsedCache.poToken,
+                        generatedAt: new Date(parsedCache.generatedAt),
+                        visitIdentifier,
+                    };
+                }
+            }
+            cache = parsedCaches;
```
Comment (grqz) [inline server/src/generate_once.ts:36]:
```suggestion
```
you're writing directly to `cache`(line 29), wouldn't this break it?

File: server/src/generate_once.ts [hunk 8]
```diff
@@ -14,37 +17,56 @@ const options = program.opts();
...
-        if (verbose) {
-            console.log(`Generated visitor data: ${generatedVisitorData}`);
-        }
-
+        if (!generatedVisitorData) process.exit(1);
+        log(`Generated visitor data: ${generatedVisitorData}`);
         visitorIdentifier = generatedVisitorData;
     }
 
     const sessionData = await sessionManager.generatePoToken(visitorIdentifier);
```
Comment (grqz) [inline server/src/generate_once.ts:62]: `visitIdentifier` or `visitorIdentifier`?

**Review by Brainicism — APPROVED** [review]

File: plugin/yt_dlp_plugins/extractor/getpot_bgutil_script.py [hunk 9]
```diff
@@ -70,10 +70,10 @@ def _get_pot(self, client: str, ydl: YoutubeDL, visitor_data=None, data_sync_id=
...
 
-        self._logger.debug(f'stdout = {stdout}')
+        self._logger.debug(f'stdout:\n{stdout}\nstderr:\n{stderr}')
         if returncode:
             raise RequestError(
-                f'_get_pot_via_script failed with returncode {returncode}:\n{stderr.strip()}')
+                f'_get_pot_via_script failed with returncode {returncode}')
 
         # The JSON response is always the last line
         script_data_resp = stdout.splitlines()[-1]
```
Comment (grqz) [inline plugin/yt_dlp_plugins/extractor/getpot_bgutil_script.py:79]: Line 80: maybe more useful if we log the `json.loads`'ed response?

Comment (Brainicism) [reply plugin/yt_dlp_plugins/extractor/getpot_bgutil_script.py:79]: is there any big difference? printing the raw json string vs however python stringifies its dictionary objects?

Comment (grqz) [reply plugin/yt_dlp_plugins/extractor/getpot_bgutil_script.py]: the raw json string should be in the stdout above already.
an example showing the difference:
```log
[debug] [GetPOT] BgUtilScriptPot: stdout:
{"visitIdentifier":"CgtaMU1xN3lTbUdSZyjcmpm3BjIKCgJBVRIEGgAgFg%3D%3D","poToken":"MniMfshIJMIyQoxCtYapTA1Wt1oxnn4fsxffqVA25ugaXNhtVU4-Teb5zXb9c4QVsM9TYJWUxddAaWTqU6aqswTF1PxcuCI-qb-AxdipvSotaB8s4uLLe3aW8hsP9R6smLQGLTmSSJ05cF8qwop-w2pGIUCcNn3nb8g=","generatedAt":"2024-09-15T02:58:39.194Z"}
[debug] [GetPOT] BgUtilScriptPot: _get_pot_via_script response = {'visitIdentifier': 'CgtaMU1xN3lTbUdSZyjcmpm3BjIKCgJBVRIEGgAgFg%3D%3D', 'poToken': 'MniMfshIJMIyQoxCtYapTA1Wt1oxnn4fsxffqVA25ugaXNhtVU4-Teb5zXb9c4QVsM9TYJWUxddAaWTqU6aqswTF1PxcuCI-qb-AxdipvSotaB8s4uLLe3aW8hsP9R6smLQGLTmSSJ05cF8qwop-w2pGIUCcNn3nb8g=', 'generatedAt': '2024-09-15T02:58:39.194Z'}
```
the differences arent big, though

**Pull request merged**